radix-engine/src/model/resources/executables/resource_manager.rs

```diff
@@ -386,7 +386,11 @@ impl Executor for ResourceManagerCreateNonFungibleInvocation {
     where
         Y: SystemApi,
     {
-        let global_node_id = api.allocate_node_id(RENodeType::GlobalResourceManager)?;
+        let global_node_id = if let Some(address) = self.resource_address {
```

Comment:
The `IdAllocator` keeps track of "allocated" `RENodeId`s. Scrypto package won't be able to allocate any RENodeId execept for it's own use.

`AuthRule` I think is the wrong abstraction to be preventing this access.

Reply:
Maybe I'm misunderstanding - but what protects against user scrypto calling `ResourceManagerCreateNonFungibleInvocation` with a fixed address?

Reply:
I'm worried about this possibility because if user-generated addresses aren't pseudorandom, it opens up the possibility for more impersonation and easy attacks on our merkle tree assumptions.

Reply:
Right, a user cannot just use an arbitrary address. To use an address a user must (pseudocode):
```
  let id = api.allocate_id();
  api.create_node(id, ..);
```
If they use a random `id` which didn't come from allocation then the `create_node` call will fail.

Reply:
Ahh okay. And it's this allocating of the id which is protected somehow?

Could we maybe add a comment explaining the above to each of these methods?

Reply:
`allocate_id` is not protected, it returns a useable id that user may use. To use a "vanity address", the address must be part of `pre_allocated_ids` in `Executable` (which is always empty for user transactions) and one may then just call:

```
api.create_node(pre_allocated_id, ..);
```

Reply:
Okay, so in that branch in the below code, what happens if the user hasn't already allocated a global node with that address - at what point will things fail?

Will it throw at `api.create_node(global_node_id, ...)`?

This is rather non-obvious for me just reviewing this, could we add a comment somewhere?